DISP-150 Fix Belgian eID verification

adessy opened the pull request:
[DISP-150](https://citizenlab.atlassian.net/browse/DISP-150)

adessy committed:
DISP-150 Fix the issuer

After some debugging, I found out that the expected issuer when verifying the access token did not match the issuer specified in the configuration.

adessy committed:
DISP-150 Add BOSA FAS public keys

Add public keys explicitly to prevent `omniauth_openid_connect` to fallback to discovery mode.

The discovery mode fails because `omniauth_openid_connect` uses the issuer URI to infer the location of the .well-known endpoint (which returns the discovery document), but the endpoint is actually located at: https://iamapps-public.int.belgium.be/.well-known/openid-configuration.

adessy commented:
@nTraum Could you have a quick look at this one? I saw you had to deal with this nasty bug in the past 🙂

adessy committed:
DISP-150 Merge branch 'master' into DISP-150/belgian-eID-verification-fails

Commit:
Translations updated by CI (extract-intl)

nTraum commented on back/engines/commercial/id_bosa_fas/app/lib/id_bosa_fas/bosa_fas_omniauth.rb:
```diff
@@ -31,10 +31,11 @@ def omniauth_setup(configuration, env)
       options[:acr_values] = 'urn:be:fedict:iam:fas:Level450'
       options[:send_scope_to_token_endpoint] = false
       options[:client_signing_alg] = :RS256
+      options[:client_jwk_signing_key] = File.read(::IdBosaFas::Engine.root.join('config', 'bosa_fas_jwks.json'))
```
Were you able to test that this works? I think I attempted something similar, by providing the jwks_uri on line 49 that links to the same JSON file:

```
   ENVIRONMENTS = {
      'integration' => {
        host: 'idp.iamfas.int.belgium.be',
        jwks_uri: 'https://idp.iamfas.int.belgium.be/fas/oauth2/connect/jwk_uri'
      },
      'production' => {
        host: 'idp.iamfas.belgium.be',
        jwks_uri: 'https://idp.iamfas.belgium.be/fas/oauth2/connect/jwk_uri'
      }
    }
```

Also, this will most likely break for the integration environment, as they use a different JSON file for it.

nTraum replied:
It might be that simply fixing the `issuer` might already fix the issue, but I cannot try that out myself (because the existing `jwks_uri` option links to the very same JSON file already, but I'm guessing here).

adessy replied:
> Also, this will most likely break for the integration environment, as they use a different JSON file for it.

Indeed, very good point! Gonna add the keys for the other environment.

> Were you able to test that this works?

Yep, I managed to test it locally. It was a bit tricky to setup though and I had to hack a few things to make it work, so the test may not be 100% reliable.

Strangely enough I could not find any usage of the `jwks_uri` in `omniauth_openid_connect`. I came to the conclusion it's just not being used by the library.

> It might be that simply fixing the issuer might already fix the issue 

I'm pretty sure it's not enough because I actually did things the other way around (the order of the commits is reversed). I first added the public keys. Then the error message changed to complain that the access token was not  valid due to a mismatch between issuers.

adessy committed:
DISP-150 Remove `jwks_uri` from client options for BOSA FAS

`jwks_uri` option does not seem to be used here. I couldn't find any direct use of it in `omniauth_openid_connect` and the following github comment suggest that it's only used in discovery mode: https://github.com/omniauth/omniauth_openid_connect/issues/118#issuecomment-1178041723

Specifying the keys directly via the client_jwk_signing_key option should suffice.

adessy committed:
DISP-150 Get JWKS for BOSA FAS from URI

Get JWKS by querying the jwks_uri endpoint instead of reading a local copy of the keys. This will protect us from key rotations at the cost of an additional query and dependence on the availability of the endpoint.

adessy committed:
DISP-150 Merge remote-tracking branch 'origin/DISP-150/belgian-eID-verification-fails' into DISP-150/belgian-eID-verification-fails

adessy committed:
DISP-150 Fix rubocop offenses

adessy committed:
DISP-150 Merge branch 'master' into DISP-150/belgian-eID-verification-fails

brentguf commented:
@adessy @nTraum is it possible to add tests for this to make sure it doesn't happen again in the future?

adessy commented:
> @adessy @nTraum is it possible to add tests for this to make sure it doesn't happen again in the future?

I'm afraid this would be pretty difficullt to implement as the test would have to interact with an external service (mocking the service would make the test useless here) and CSAM / BOSA FAS don't provide test accounts to the best of my knowledge. It could maybe be partially covered by an end-to-end test.

brentguf commented:
> It could maybe be partially covered by an end-to-end test.

But then we need a test account as well?
